### PR TITLE
XBee node discovery

### PR DESCRIPTION
**Pull request by timvandermeij**
This patch implements a devices view in the control panel that allows us to track the status of all XBee devices in the network. We use XBee's built-in node discovery functionality to let the ground station discover other devices in the network. When we found a device in the network, the `ND` packet contains its node identifier and address, which we store in the devices view.

We show the devices view immediately after the loading view because one needs to be sure that the XBees have joined the network before sending any waypoints to them.

@lhelwerd I already did extensive tests with this, but I do not merge this PR immediately in case you would like to look at the code too. Let me know if you have any comments or if you're fine with it, because then I'll update the PR and merge it. It's probably best to look at the overall diff and it should not be too hard to follow, but I did add some more comments in the commit messages.

### CODE REVIEW

**Comment by timvandermeij** [comment]
Since it's becoming increasingly hard to test this without the XBees, below is a screenshot of the view for you to see how it looks (click to enlarge).

device view

File: zigbee/XBee_Sensor_Simulator.py [hunk 1]
```diff
@@ -126,6 +138,16 @@ def enqueue(self, packet, to=None):
...
 
+    def discover(self, callback):
+        """
+        Discover other XBee devices in the network.
+
+        This method is only used on the ground station in the control panel
+        to refresh the status of the other XBee devices.
+        """
+
+        pass
```
Comment (lhelwerd) [inline zigbee/XBee_Sensor_Simulator.py]: I guess I cannot really test it right now with the XBee simulator ;) I'm not sure how difficult it is to implement the discovery for the simulators. Could you just send a special packet/command to all the socket ports (given we know the maximum number of sensors beforehand) and respond to that with another special command? Then it would be quite similar to the physical version.

**Comment by lhelwerd** [comment]
The code looks fine by me. I suppose you need to refresh at least once to see the sensors? I think that should work fine, although automatic detection would of course be best :wink:

The GUI takes up :100:% CPU of one core for me, but this also happens on the master branch. Perhaps we should file a separate issue for that or figure out why it happens.

**Comment by lhelwerd** [comment]
Apparently the 100% CPU usage is specific to the `--controller-xbee-simulation` flag, since it does not seem to happen in the loading view for physical XBees at least. Perhaps you can confirm. The profiler does not give conclusive results, perhaps because it does not profile all threads. It's not a big issue since it's simulator-only, but I could delve into it more.

File: zigbee/XBee_Sensor_Simulator.py [hunk 2]
```diff
@@ -87,6 +87,7 @@ def _loop(self):
                 try:
                     data = self._socket.recv(self.settings.get("buffer_size"))
                 except socket.error:
+                    time.sleep(self._loop_delay)
```
Comment (lhelwerd) [inline zigbee/XBee_Sensor_Simulator.py:90]: Nice find!

File: zigbee/XBee_Sensor_Simulator.py [hunk 3]
```diff
@@ -146,7 +146,15 @@ def discover(self, callback):
         to refresh the status of the other XBee devices.
         """
 
-        pass
+        # The simulator does not use XBee device discovery because it does not
+        # use the actual XBee library that provides this functionality. We
+        # simulate the process by calling the callback with the packet manually.
+        for vehicle in [1, 2]:
```
Comment (lhelwerd) [inline zigbee/XBee_Sensor_Simulator.py]: Perhaps you can make this more dynamic by using `for vehicle in range(1, self.scheduler.number_of_sensors+1)` instead. We can assume for now in the control panel view that we have 2 vehicle sensors, but we should probably solve this in a similar way later on as well.

**Comment by timvandermeij** [comment]
I have addressed your comment about the `discover` method in the simulator. I have also fixed the issue with 100% CPU usage. The cause of that issue was kind of orthogonal to this patch, but it's so small we can fix it here anyway. I could not reproduce the issue at first, but then thought it might be something with the simulator. Only shortly after that I noticed your comment, so you found it at the exact same moment as I did!

Indeed we have to press the refresh button at least once, but most of the time it's only once, so that's a tiny effort ;-) Making that automatic is not trivial as we might need some kind of timer mechanism, so let's not do that for now.

File: zigbee/XBee_Sensor_Simulator.py (hunk 3 above)
Comment (timvandermeij) [inline zigbee/XBee_Sensor_Simulator.py]: Good point. I have fixed that in the new commit.

**Comment by timvandermeij** [comment]
Let's merge this as it's done and the additional concerns have been addressed. Thanks for the review!

**Comment by lhelwerd** [comment]
:+1:
